tamr_client/operation.py

```diff
@@ -0,0 +1,195 @@
+"""
+See https://docs.tamr.com/new/reference/the-operation-object
+"""
+from copy import deepcopy
+from dataclasses import dataclass
+from time import sleep, time as now
+from typing import Dict, Optional
+
+import requests
+
+from tamr_client import response
+from tamr_client.instance import Instance
+from tamr_client.session import Session
+from tamr_client.types import JsonDict
+from tamr_client.url import URL
+
+
+class NotFound(Exception):
+    """Raised when referencing an operation that does not exist on the server.
+    """
+
+    pass
+
+
+@dataclass(frozen=True)
+class Operation:
+    """A Tamr operation
+
+    See https://docs.tamr.com/new/reference/the-operation-object
+
+    Args:
+        url
+    """
+
+    url: URL
+    id: str
+    type: str
+    status: Optional[Dict[str, str]] = None
+    description: Optional[str] = None
+
+
+def from_response(instance: Instance, response: requests.Response) -> Operation:
+    """
+    Handle idiosyncrasies in constructing Operations from Tamr responses.
+    When a Tamr API call would start an operation, but all results that would be
+    produced by that operation are already up-to-date, Tamr returns `HTTP 204 No Content`
+
+    To make it easy for client code to handle these API responses without checking
+    the response code, this method will either construct an Operation, or a
+    dummy `NoOp` operation representing the 204 Success response.
+
+    Args:
+        response: HTTP Response from the request that started the operation.
+    """
+    if response.status_code == 204:
+        # Operation was successful, but the response contains no content.
+        # Create a dummy operation to represent this.
+        _never = "0000-00-00T00:00:00.000Z"
+        _description = """Tamr returned HTTP 204 for this operation, indicating that all
+            results that would be produced by the operation are already up-to-date."""
+        resource_json = {
+            "id": "-1",
+            "type": "NOOP",
+            "description": _description,
+            "status": {
+                "state": "SUCCEEDED",
+                "startTime": _never,
+                "endTime": _never,
+                "message": "",
+            },
+            "created": {"username": "", "time": _never, "version": "-1"},
+            "lastModified": {"username": "", "time": _never, "version": "-1"},
+            "relativeId": "operations/-1",
+        }
+    else:
+        resource_json = response.json()
+    _id = resource_json["id"]
+    _url = URL(instance=instance, path=f"operations/{_id}")
+    return _from_json(_url, resource_json)
+
+
+def apply_options(
```

Comment:
same comment about `_` prefix:

```suggestion
def _apply_options(
```

This would let us remove the `**NOTE**` in the docstring too.

Reply:
There's a design question here about whether to include this function at all, or if users should just call `wait` themselves explicitly.

Additionally, it would be useful to ask the field how often, if ever, they use `asynchronous` mode.

Reply:
It looks like `apply_options` is used internally in TUC in a few places, mostly `refresh` methods for datasets.  It seems like this is used just to pass along the details of polling frequency and client timeouts.